tests/test_browser.py

```diff
@@ -4553,3 +4553,12 @@ def test_no_declare_asm_module_exports_asmjs(self):
   @no_wasm_backend('MINIMAL_RUNTIME not yet available in Wasm backend')
   def test_no_declare_asm_module_exports_wasm_minimal_runtime(self):
     self.btest(path_from_root('tests', 'declare_asm_module_exports.cpp'), '1', args=['-s', 'DECLARE_ASM_MODULE_EXPORTS=0', '-s', 'ENVIRONMENT=web', '-O3', '--closure', '1', '-s', 'MINIMAL_RUNTIME=1'])
+
+  # Tests that the different code paths in src/shell_minimal_runtime.html all work ok.
+  @no_wasm_backend('MINIMAL_RUNTIME not yet available in Wasm backend')
+  def test_minimal_runtime_loader_shell(self):
+    args = ['-s', 'MINIMAL_RUNTIME=2']
+    for wasm in [[], ['-s', 'WASM=0', '--memory-init-file', '0'], ['-s', 'WASM=0', '--memory-init-file', '1']]:
+      for modularize in [[], ['-s', 'MODULARIZE=1']]:
+        print(str(args + wasm + modularize))
+        self.btest('minimal_hello.c', '0', args=args + wasm + modularize)
```

Comment:
How does this test work? MODULARIZE creates a function, but user code needs to call it?

Reply:
In other words, the default output with MODULARIZE runs no code (unlike MODULARIZE_INSTANCE), so how is anything running here?

Reply:
Oh, I modeled this after the previous understanding of what MODULARIZE=1 does.

In my view, MODULARIZE just wraps the generated JS code into a function, instead of leaving it to litter global scope.

The decision when to instantiate the function is then on the shoulders of the `--shell-file` that is passed. The default behavior for the bundled shell file is to showcase how this instantiation is done, which for this specific case is done on this line: https://github.com/emscripten-core/emscripten/pull/8057/files#diff-2e68559aa6b479fdda40e43a5a6a5b6aR35

If one wants to adjust that instantiation to not occur, then one would provide their custom --shell-file with another one that instantiates when they want it to happen, or doesn't.

It is odd to have the default provided shell not run the program. It is kind of "with MODULARIZE, you will need to manually instantiate the code, but our example shell file will not show you how that is supposed to happen". (This is about the same as the other conversation around MODULARIZE+USE_PTHREADS PR: why have partial/unfunctional example shell files in the repository at all?)

Default shell file is just that, a default. It should showcase a complete scenario of how one builds an application, having the default shell file not do that makes it look like the default shell file is lacking/buggy/missing needed instructions.

Perhaps there is a confusion about what exactly controls the instantiation here?

If you do
`-s MODULARIZE=1 -o a.js`, you will get a JS file that does not launch automatically. If you do
`-s MODULARIZE=1 -o a.html --shell-file myshell.html`, you will likewise get a JS file that does not launch automatically. The html file will neither launch automatically, it only does so if you wrote myshell.html to instantiate the function yourself.
If you do
`-s MODULARIZE=1 -o a.html`, you will get a JS file that will not launch automatically. Since you did not specify a shell file, you will get the default shell file that has the default behavior of launching automatically.

There does not seem to exist an use case with anyone wanting/expecting to get `-s MODULARIZE=1 -o a.html` to generate an a.html file from default shell that does not do anything. Who would want that? It just seems broken and useless behavior if default shell files should be nonfunctional under MODULARIZE=1?

One might argue that their use case was to do `-s MODULARIZE=1 -o a.html --post-js thisJsFileShouldLaunchThePageAndNotTheDefaultHtmlShellFile.js` and they really insist on being able to use the default shell html file for that purpose and not have to provide their own. But that is a really silly expectation (just copy the default shell file to your own project, and comment out the instantiation line from there if you really need that)

Reply:
It seems like your suggested change makes MODULARIZE behave like MODULARIZE_INSTANCE in the specific case where the user uses the default HTML shell. I get your point that with the default HTML, where could the code be instantiated? But on the other hand, that makes the MODULARIZE / MODULARIZE_INSTANCE difference less clear.

I think the default HTML is not that important - most users will replace it anyhow. For that reason I don't feel terribly strongly about this.

But if we want to consider this change, we'd need to make it in non-minimal as well, in a separate PR that notes this breaking change, documents it consistently, etc. If so let's open an issue or PR to discuss that and cc the people that helped design that feature.

Reply:
> It seems like your suggested change makes MODULARIZE behave like MODULARIZE_INSTANCE in the specific case where the user uses the default HTML shell.

The intent is not to make MODULARIZE behave like MODULARIZE_INSTANCE. `MODULARIZE_INSTANCE` has a special kind of polymorphic behavior where loading the JS code starts off with `Module` as a function, but then it self-invokes and replaces itself with an exports object, destroying the original `Module` function. That means `MODULARIZE_INSTANCE` disallows one from instantiating the JS code multiple times, and it disallows one to choose when/where to instantiate.

Also, this is not changing how `MODULARIZE` option itself behaves, but this is just ensuring that the default shell also works when it is used in a modularized build.

> most users will replace it anyhow
> a separate PR that notes this breaking change

This is not a breaking change, or even a behavior adjustment to how `MODULARIZE` itself works. Rather, this PR just establishes how the default shell for MINIMAL_RUNTIME operates with MODULARIZE. ("establishes", meaning, "make the default shell work with MODULARIZE without having to replace the default shell")

Also, (sorry, something I forgot while writing previous reply, but now remember after re-reading the earlier conversation about MODULARIZE+PTHREADS) this is not a breaking change for default shell for non-MINIMAL_RUNTIME, because our current default shell is broken&unusable with MODULARIZE. This is because our default shell creates a global `var Module = {};` imports property object in it, but then since default value of `-s EXPORT_NAME` setting is also `Module`, the generated `a.js` file will create a global `function Module()` object. Upon loading `a.js`, that function will overwrite `var Module` from the default shell html, so the imports property object is lost. One would have to type in console effectively `Module(Module);` to start, where first `Module` is the module function from `a.js`, and the parameter is the `Module` property from `a.html`, but those collided in name. So currently if you are using MODULARIZE, you cannot use the default shell, unless you renamed EXPORT_NAME to something else to avoid collision.

Then when you do the rename with something like `-s MODULARIZE=1 -s EXPORT_NAME=EmscriptenCode`, you still get a html page that is not functional, but you have to manually open the console and type in `EmscriptenCode(Module);` to run the content. That does not sound like anyone's use case.

> we'd need to make it in non-minimal as well,

In the earlier PR about MODULARIZE+PTHREADS, I did already propose a fix to the default shell that avoided the name collision on `Module`, by changing EXPORT_NAME to `EmscriptenCode` when one uses the default shell (otherwise it'd still be `Module`), and when default shell is used, emit in the generated `a.html` a call to `EmscriptenCode(Module);` so that the page does run. This rename from `Module` to `EmscriptenCode` would likely not be visible to user since they are not using their own HTML shell file.

Another fix might be to change default shell to use some other variable name, like `var emscriptenImports = {};` for the config object in `MODULARIZE` mode, and in non-MODULARIZE mode, it'd still use `Module`. Or something else. The root problem is overloading the `Module` name for different purposes in different modes, and the reason why this is easier for MINIMAL_RUNTIME to avoid is that it removes uses the concepts of `Module`.

I have a bit of fatigue to attempt to go and revisit to wrangle the default shell to be compatible with MODULARIZE now, because it requires fixing up the `Module`+`Module` name conflict one way or the other, and the way I attempted to do it was turned down once already (I don't say that was wrong call to do, but that clash needs to be addressed).

Reply:
Thanks for the details.

This all makes me wonder - is it worth supporting html output with MODULARIZE? As you say, it didn't work - and no one complained. To be honest, I think it wasn't tested because I and other people that thought about MODULARIZE assumed it would *not* be used as a complete app, that is, it was just a JS script that some app with it's own HTML would include.

Anyhow, I don't feel strongly about this, given the points you make. I'm ok with either

 * fix MODULARIZE with the default shell, as you do here
 * disallow MODULARIZE with HTML output

but in both cases, let's make the minimal and non-minimal cases behave the same. If we fix it as you propose, please mention in the MODULARIZE docs in settings.js that with HTML output we do actually run the code, as otherwise nothing else would.

Reply:
After investigating the `script.un_src()` and `script.inline` logic in emcc.py, unfortunately I cannot make minimal runtime and traditional runtime behave the same here. It is too complex and too time consuming to try to refactor traditional runtime.

The best I can do here is to add an error abort to compilation frontend when default shell is used with MODULARIZE=1 in traditional runtime, and add a note about the default shell being incompatible with MODULARIZE=1, but the default shell in MINIMAL_RUNTIME works with MODULARIZE=1. Updated the PR with those. How does this look now?

Reply:
Yeah, sounds reasonable to error on that case.